core/indigo-core/reaction/src/reaction_multistep_detector.cpp

```diff
@@ -238,15 +234,470 @@ void ReactionMultistepDetector::createSummBlocks()
+                _components[cluster[k]].idx = (int)i;
+                _components[cluster[k]].mol.reset();
+            }
+        }
+    }
+
+    for (int i = 0; i < _moleculeCount; ++i)
+    {
+        if (_components[i].mol)
+        {
+            _merged_components.emplace_back(std::move(_components[i].mol), _components[i].hull, (int)_merged_components.size());
+            _merged_components.back().mapped_idx = i;
+            _components[i].mapped_idx = (int)_merged_components.size() - 1;
+        }
+    }
+    _moleculeCount = (int)_merged_components.size();
+}
+
+void dumpHull(std::string name, const std::vector<Vec2f>& hull)
+{
+    std::cout << name << std::endl;
+    for (auto& v : hull)
+        std::cout << v.x << "," << v.y << std::endl;
+}
+
+std::optional<std::pair<int, int>> ReactionMultistepDetector::isMergeable(size_t mol_idx1, size_t mol_idx2, std::optional<std::pair<int, int>>& current_zone)
+{
+    auto& mdi = _mol_distances[mol_idx1];
+    auto dist_it = mdi.distances_map.find(mol_idx2);
+    if (dist_it != mdi.distances_map.end() && dist_it->second < LayoutOptions::DEFAULT_BOND_LENGTH * 2)
+    {
+        // collect surrounding zones for both molecules
+        std::map<int, std::set<int>> other_zones1;
+        std::map<int, std::set<int>> other_zones2;
+        std::map<int, std::set<int>> comm_zones;
+
+        auto zone1 = findMaxSpecialZone(mol_idx1, other_zones1);
+        auto zone2 = findMaxSpecialZone(mol_idx2, other_zones2);
+        std::set_intersection(other_zones1.begin(), other_zones1.end(), other_zones2.begin(), other_zones2.end(), std::inserter(comm_zones, comm_zones.begin()),
+                              [](auto& a, auto& b) { return a.first < b.first; });
+
+        // no zone - no reason to merge
+        if (zone1.has_value())
+        {
+            // if both molecules has the same zone and section - merge
+            // we never merge molecules with different sections of the same zone
+            if (zone2.has_value())
+            {
+                // disable catalysts merging
+                if ((_zones[zone1.value().first].zone_type == ZoneType::EArrow && zone1.value().second > 1) ||
+                    (_zones[zone2.value().first].zone_type == ZoneType::EArrow && zone2.value().second > 1))
+                    return std::nullopt;
+
+                // if both molecules are on the same zone - check if they are mergeable
+
+                // check if there are opposite sections
+                for (auto& cz : comm_zones)
+                {
+                    auto it_oz2 = other_zones2.find(cz.first);
+                    if (it_oz2 != other_zones2.end())
+                    {
+                        for (auto& section : cz.second)
+                        {
+                            if ((_zones[cz.first].zone_type == ZoneType::EPathWay && section > 1) || it_oz2->second.count(section ^ 1))
```

Comment:
It's not clear what means `it_oz2->second.count(section ^ 1)`

Reply:
Added clarifying comment

Reply:
According to `findMaxSpecialZone` `zone1`, `zone2` and `comm_zone` are `std::map<int, std::set<int>>` where key is index in `_zones` and `std::set<int>` is indexes of sections form this zone which intersect with molecule.
So, it's works only why sections inserted into zones in current order(I'm not sure about pathway zone which has one right zone and from 0 to n left zones)
If later section adding will be changed - this part of code broke.
Same in lines 824-825:
```C++
                            auto& top_zone = _zones[reaction_index].zone_sections[2];
                            auto& bottom_zone = _zones[reaction_index].zone_sections[3];
```
Maybe better define sections explicitly?